File: src/java/org/orbeon/oxf/portlet/Portlet2ExternalContext.java
/**
 * Copyright (C) 2010 Orbeon, Inc.
 *
 * This program is free software; you can redistribute it and/or modify it under the terms of the
 * GNU Lesser General Public License as published by the Free Software Foundation; either version
 * 2.1 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * The full text of the license is available at http://www.gnu.org/copyleft/lesser.html
 */
package org.orbeon.oxf.portlet;

import org.orbeon.oxf.common.OXFException;
import org.orbeon.oxf.externalcontext.PortletToExternalContextRequestDispatcherWrapper;
import org.orbeon.oxf.externalcontext.URLRewriter;
import org.orbeon.oxf.externalcontext.WSRPURLRewriter;
import org.orbeon.oxf.pipeline.api.ExternalContext;
import org.orbeon.oxf.pipeline.api.PipelineContext;
import org.orbeon.oxf.processor.serializer.CachedSerializer;
import org.orbeon.oxf.servlet.ServletExternalContext;
import org.orbeon.oxf.util.*;

import javax.portlet.*;
import java.io.*;
import java.security.Principal;
import java.util.*;

/*
 * Portlet-specific implementation of ExternalContext.
 */
public class Portlet2ExternalContext extends PortletWebAppExternalContext implements ExternalContext {

    private static final String OPS_CONTEXT_NAMESPACE_KEY = "org.orbeon.ops.portlet.namespace";

    private static RequestFilter requestFilter;
    static {
        try {
            final Class<? extends RequestFilter> customContextClass
                    = (Class<? extends RequestFilter>) Class.forName("org.orbeon.oxf.portlet.FormRunnerRequestFilter");
            requestFilter = customContextClass.newInstance();
        } catch (Exception e) {
            // Silently ignore as this typically means that we are not in Liferay
        }
    }

    private ExternalContext.Request request;
    private ExternalContext.Response response;
    private ExternalContext.Session session;
    private ExternalContext.Application application;

    private PipelineContext pipelineContext;
    private PortletRequest portletRequest;
    private ClientDataRequest clientDataRequest;
    private MimeResponse mimeResponse;

    Portlet2ExternalContext(PipelineContext pipelineContext, PortletContext portletContext, Map<String, String> initAttributesMap, PortletRequest portletRequest) {
        this(portletContext, initAttributesMap);
        this.pipelineContext = pipelineContext;

        // Wrap request if needed
        if (requestFilter != null) {
            try {
                this.portletRequest = requestFilter.amendRequest(portletRequest);
            } catch (Exception e) {
                throw new OXFException(e);
            }
        } else {
            this.portletRequest = portletRequest;
        }

        if (portletRequest instanceof ClientDataRequest)
            this.clientDataRequest = (ClientDataRequest) portletRequest;
    }

    Portlet2ExternalContext(PipelineContext pipelineContext, PortletContext portletContext, Map<String, String> initAttributesMap, PortletRequest portletRequest, MimeResponse mimeResponse) {
        this(pipelineContext, portletContext, initAttributesMap, portletRequest);
        this.mimeResponse = mimeResponse;
    }

    private Portlet2ExternalContext(PortletContext portletContext, Map<String, String> initAttributesMap) {
        super(portletContext, initAttributesMap);
    }

    private class Request implements ExternalContext.Request {

        private Request() {}

        private String namespace = null;

        private Map<String, Object> attributesMap;
        private Map<String, String[]> headerValuesMap;
        private Map<String, Object[]> parameterMap;

        public String getContainerType() {
            return "portlet";
        }

        public String getContainerNamespace() {
            return getResponse().getNamespacePrefix();
        }

        public String getContextPath() {
            return portletRequest.getContextPath();
        }

        public String getPathInfo() {
            // Use the resource id if we are a ResourceRequest
            // In that case, remove the query string part of the resource id, that's handled by getParameterMap()
            String result = (portletRequest instanceof ResourceRequest) ? NetUtils.removeQueryString(((ResourceRequest) portletRequest).getResourceID()) : portletRequest.getParameter(OrbeonPortletXFormsFilter.PATH_PARAMETER_NAME);
            if (result == null) result = "";
            return (result.startsWith("/")) ? result : "/" + result;
        }

        public String getRemoteAddr() {
            // NOTE: The portlet API does not provide for this value.
            return null;
        }

        public synchronized Map<String, String[]> getHeaderValuesMap() {
            // NOTE: The container may or may not make HTTP headers available through the properties API
            if (headerValuesMap == null) {
                headerValuesMap = new HashMap<String, String[]>();

                // NOTE: Not sure we should even pass these properties as "headers"
                // Example of property: javax.portlet.markup.head.element.support = true
                for (Enumeration<String> e = portletRequest.getPropertyNames(); e.hasMoreElements();) {
                    String name = e.nextElement();
                    // NOTE: Normalize names to lowercase to ensure consistency between servlet containers
                    headerValuesMap.put(name.toLowerCase(), StringConversions.stringEnumerationToArray(portletRequest.getProperties(name)));
                }

                // PLT.11.1.5 Request Properties: "client request HTTP headers may not be always available. Portlets
                // should not rely on the presence of headers to function properly. The PortletRequest interface
                // provides specific methods to access information normally available as HTTP headers: content-length,
                // content-type, accept-language."
                // NOTE: It seems like while Liferay 5 was making headers available, Liferay 6 doesn't anymore.
                if (portletRequest instanceof ClientDataRequest) {
                    final ClientDataRequest clientDataRequest = (ClientDataRequest) portletRequest;
                    if (clientDataRequest.getContentType() != null)
                        headerValuesMap.put("content-type", new String[] { clientDataRequest.getContentType() });
                    if (clientDataRequest.getContentLength() != -1)
                        headerValuesMap.put("content-length", new String[] { Integer.toString(clientDataRequest.getContentLength()) });
                }
            }
            return headerValuesMap;
        }

        public ExternalContext.Session getSession(boolean create) {
            return Portlet2ExternalContext.this.getSession(create);
        }

        public void sessionInvalidate() {
            PortletSession session = portletRequest.getPortletSession(false);
            if (session != null)
                session.invalidate();
        }

        public String getAuthType() {
            return portletRequest.getAuthType();
        }

        public String getRemoteUser() {
            return portletRequest.getRemoteUser();
        }

        public boolean isSecure() {
            return portletRequest.isSecure();
        }

        public boolean isUserInRole(String role) {
            return portletRequest.isUserInRole(role);
        }

        public String getCharacterEncoding() {
            return (clientDataRequest != null) ? clientDataRequest.getCharacterEncoding() : null;
        }

        public int getContentLength() {
            return (clientDataRequest != null) ? clientDataRequest.getContentLength() : -1;
        }

        public String getContentType() {
            return (clientDataRequest != null) ? clientDataRequest.getContentType() : null;
        }

        public String getMethod() {
            return (clientDataRequest != null) ? clientDataRequest.getMethod() : null;
        }

        public synchronized Map<String, Object[]> getParameterMap() {
            if (parameterMap == null) {
                // Two conditions: file upload ("multipart/form-data") or not
                if (getContentType() != null && getContentType().startsWith("multipart/form-data")) {
                    // Special handling for multipart/form-data

                    // Decode the multipart data
                    parameterMap = Multipart.getParameterMapMultipart(pipelineContext, request, ServletExternalContext.DEFAULT_FORM_CHARSET_DEFAULT);
                } else if (portletRequest instanceof ResourceRequest) {
                    // We encoded query parameters directly into the resource id in this case
                    final String queryString = NetUtils.getQueryString(((ResourceRequest) portletRequest).getResourceID());
                    if (queryString != null)
                        parameterMap = Collections.unmodifiableMap(StringConversions.stringArrayMapToObjectArrayMap(NetUtils.decodeQueryString(queryString, false)));
                    else
                        parameterMap = Collections.emptyMap();
                } else {
                    // Not a resource request, so just use native request parameters
                    parameterMap = new HashMap<String, Object[]>(portletRequest.getParameterMap());
                    parameterMap.remove(OrbeonPortletXFormsFilter.PATH_PARAMETER_NAME);
                    parameterMap = Collections.unmodifiableMap(parameterMap);
                }
            }
            return parameterMap;
        }

        public synchronized Map<String, Object> getAttributesMap() {
            if (attributesMap == null) {
                attributesMap = new Portlet2ExternalContext.RequestMap(portletRequest);
            }
            return attributesMap;
        }

        public String getPathTranslated() {
            return null;
        }

        public String getProtocol() {
            // NOTE: The portlet API does not provide for this value.
            return null;
        }

        public String getQueryString() {
            // NOTE: We could build one from the parameters
            return null;
        }

        public Reader getReader() throws IOException {
            return (clientDataRequest != null) ? clientDataRequest.getReader() : null;
        }

        public InputStream getInputStream() throws IOException {
            return (clientDataRequest != null) ? clientDataRequest.getPortletInputStream() : null;
        }

        public String getRemoteHost() {
            // NOTE: The portlet API does not provide for this value.
            return null;
        }

        public String getRequestedSessionId() {
            return portletRequest.getRequestedSessionId();
        }

        public String getRequestPath() {
            return getPathInfo();
        }

        public String getRequestURI() {
            // NOTE: The portlet API does not provide for this value.
            return null;
        }

        public String getRequestURL() {
            // NOTE: The portlet API does not provide for this value.
            return null;
        }

        public String getScheme() {
            return portletRequest.getScheme();
        }

        public String getServerName() {
            return portletRequest.getServerName();
        }

        public int getServerPort() {
            return portletRequest.getServerPort();
        }

        public String getServletPath() {
            // NOTE: The portlet API does not provide for this value.
            return null;
        }

        private String platformClientContextPath;
        private String applicationClientContextPath;

        public String getClientContextPath(String urlString) {
            // Return depending on whether passed URL is a platform URL or not
            return URLRewriterUtils.isPlatformPath(urlString) ? getPlatformClientContextPath() : getApplicationClientContextPath();
        }

        private String getPlatformClientContextPath() {
            if (platformClientContextPath == null) {
                platformClientContextPath = URLRewriterUtils.getClientContextPath(this, true);
            }
            return platformClientContextPath;
        }

        private String getApplicationClientContextPath() {
            if (applicationClientContextPath == null) {
                applicationClientContextPath = URLRewriterUtils.getClientContextPath(this, false);
            }
            return applicationClientContextPath;
        }

        public Locale getLocale() {
            return portletRequest.getLocale();
        }

        public Enumeration getLocales() {
            return portletRequest.getLocales();
        }

        public boolean isRequestedSessionIdValid() {
            return portletRequest.isRequestedSessionIdValid();
        }

        public Principal getUserPrincipal() {
            return portletRequest.getUserPrincipal();
        }

        public Object getNativeRequest() {
            return Portlet2ExternalContext.this.getNativeRequest();
        }

        public String getPortletMode() {
            return portletRequest.getPortletMode().toString();
        }

        public String getWindowState() {
            return portletRequest.getWindowState().toString();
        }
    }

    private class Session implements ExternalContext.Session {

        private PortletSession portletSession;
        private Map<String, Object>[] sessionAttributesMaps;

        public Session(PortletSession httpSession) {
            this.portletSession = httpSession;
        }

        public long getCreationTime() {
            return portletSession.getCreationTime();
        }

        public String getId() {
            return portletSession.getId();
        }

        public long getLastAccessedTime() {
            return portletSession.getLastAccessedTime();
        }

        public int getMaxInactiveInterval() {
            return portletSession.getMaxInactiveInterval();
        }

        public void invalidate() {
            portletSession.invalidate();
        }

        public boolean isNew() {
            return portletSession.isNew();
        }

        public void setMaxInactiveInterval(int interval) {
            portletSession.setMaxInactiveInterval(interval);
        }

        public Map<String, Object> getAttributesMap() {
            return getAttributesMap(PortletSession.PORTLET_SCOPE);
        }

        public Map<String, Object> getAttributesMap(int scope) {
            if (sessionAttributesMaps == null) {
                sessionAttributesMaps = new Map[2];
            }
            if (sessionAttributesMaps[scope - 1] == null) {
                PortletSession session = portletRequest.getPortletSession(false);
                if (session != null)
                    sessionAttributesMaps[scope - 1] = new SessionMap(session, scope);
            }
            return sessionAttributesMaps[scope - 1];
        }

        public void addListener(SessionListener sessionListener) {
            ServletExternalContext.SessionListeners listeners = (ServletExternalContext.SessionListeners) portletSession.getAttribute(ServletExternalContext.SESSION_LISTENERS, PortletSession.APPLICATION_SCOPE);
            if (listeners == null) {
                listeners = new ServletExternalContext.SessionListeners();
                portletSession.setAttribute(ServletExternalContext.SESSION_LISTENERS, listeners, PortletSession.APPLICATION_SCOPE);
            }
            listeners.addListener(sessionListener);
        }

        public void removeListener(SessionListener sessionListener) {
            final ServletExternalContext.SessionListeners listeners = (ServletExternalContext.SessionListeners) portletSession.getAttribute(ServletExternalContext.SESSION_LISTENERS, PortletSession.APPLICATION_SCOPE);
            if (listeners != null)
                listeners.removeListener(sessionListener);
        }
    }


    private class Application implements ExternalContext.Application {
        private PortletContext portletContext;

        public Application(PortletContext portletContext) {
          this.portletContext = portletContext;
        }

        public void addListener(ApplicationListener applicationListener) {

            ServletExternalContext.ApplicationListeners listeners = (ServletExternalContext.ApplicationListeners) portletContext.getAttribute(ServletExternalContext.APPLICATION_LISTENERS);
            if (listeners == null) {
                listeners = new ServletExternalContext.ApplicationListeners();
                portletContext.setAttribute(ServletExternalContext.APPLICATION_LISTENERS, listeners);
            }
            listeners.addListener(applicationListener);
        }

        public void removeListener(ApplicationListener applicationListener) {
            final ServletExternalContext.ApplicationListeners listeners = (ServletExternalContext.ApplicationListeners) portletContext.getAttribute(ServletExternalContext.APPLICATION_LISTENERS);
            if (listeners != null)
                listeners.removeListener(applicationListener);
        }
    }

    public abstract static class BaseResponse implements ExternalContext.Response {

        private final URLRewriter urlRewriter;

        public BaseResponse(PipelineContext pipelineContext, ExternalContext.Request request) {
            urlRewriter = new WSRPURLRewriter(pipelineContext, request, URLRewriterUtils.isWSRPEncodeResources());
        }

        public boolean checkIfModifiedSince(long lastModified, boolean allowOverride) {
            // NIY / FIXME
            return true;
        }

        public void setContentLength(int len) {
            // NOP
        }

        public void sendError(int code) throws IOException {
            // FIXME/XXX: What to do here?
            throw new OXFException("Error while processing request: " + code);
        }

        public void setCaching(long lastModified, boolean revalidate, boolean allowOverride) {
            // NIY / FIXME
        }

        public void setResourceCaching(long lastModified, long expires) {
            // NIY / FIXME
        }

        public void setHeader(String name, String value) {
            // NIY / FIXME: This may not make sense here. Used only by XLS serializer as of 7/29/03.
        }

        public void addHeader(String name, String value) {
            // NIY / FIXME: This may not make sense here. Used only by XLS serializer as of 7/29/03.
        }

        public void setStatus(int status) {
            // Test error
            if (status == SC_NOT_FOUND) {
                throw new OXFException("Resource not found. HTTP status: " + status);
            } else if (status >= 400) {
                // Ignore
            }
            // FIXME: How to handle NOT_MODIFIED?
        }

        public String rewriteActionURL(String urlString) {
            return urlRewriter.rewriteActionURL(urlString);
        }

        public String rewriteRenderURL(String urlString) {
            return urlRewriter.rewriteRenderURL(urlString);
        }

        public String rewriteActionURL(String urlString, String portletMode, String windowState) {
            return urlRewriter.rewriteActionURL(urlString, portletMode, windowState);
        }

        public String rewriteRenderURL(String urlString, String portletMode, String windowState) {
            return urlRewriter.rewriteRenderURL(urlString, portletMode, windowState);
        }

        public String rewriteResourceURL(String urlString, boolean generateAbsoluteURL) {
            return rewriteResourceURL(urlString, generateAbsoluteURL ? ExternalContext.Response.REWRITE_MODE_ABSOLUTE : ExternalContext.Response.REWRITE_MODE_ABSOLUTE_PATH_OR_RELATIVE);
        }

        public String rewriteResourceURL(String urlString, int rewriteMode) {
            return urlRewriter.rewriteResourceURL(urlString, rewriteMode);
        }

        public String getNamespacePrefix() {
            return urlRewriter.getNamespacePrefix();
        }

        public Object getNativeResponse() {
            throw new OXFException("NIY");
//            return Portlet2ExternalContext.this.getNativeResponse();
        }
    }

    public static class BufferedResponse extends BaseResponse {

        private String contentType;
        private String redirectPathInfo;
        private Map<String, String[]> redirectParameters;
        private boolean redirectIsExitPortal;
        private StringBuilderWriter stringBuilderWriter;
        private PrintWriter printWriter;
        private ByteArrayOutputStream byteStream;
        private String title;
        
        public BufferedResponse(PipelineContext pipelineContext, ExternalContext.Request request) {
            super(pipelineContext, request);
        }
        
        public String getContentType() {
            return contentType;
        }

        public String getRedirectPathInfo() {
            return redirectPathInfo;
        }

        public Map<String, String[]> getRedirectParameters() {
            return redirectParameters;
        }

        public StringBuilderWriter getStringBuilderWriter() {
            return stringBuilderWriter;
        }

        public ByteArrayOutputStream getByteStream() {
            return byteStream;
        }

        public boolean isRedirectIsExitPortal() {
            return redirectIsExitPortal;
        }

        public PrintWriter getWriter() throws IOException {
            if (stringBuilderWriter == null) {
                stringBuilderWriter = new StringBuilderWriter();
                printWriter = new PrintWriter(stringBuilderWriter);
            }
            return printWriter;
        }

        public OutputStream getOutputStream() throws IOException {
            if (byteStream == null)
                byteStream = new ByteArrayOutputStream();
            return byteStream;
        }

        public void setContentType(String contentType) {
            this.contentType = NetUtils.getContentTypeMediaType(contentType);
        }

        /**
         *
         * @param pathInfo      path to redirect to
         * @param parameters    parameters to redirect to
         * @param isServerSide  this is ignored for portlets
         * @param isExitPortal  if this is true, the redirect will exit the portal
         */
        public void sendRedirect(String pathInfo, Map<String, String[]> parameters, boolean isServerSide, boolean isExitPortal) {
            if (isCommitted())
                throw new IllegalStateException("Cannot call sendRedirect if response is already committed.");
            this.redirectPathInfo = pathInfo;
            this.redirectParameters = parameters;
            this.redirectIsExitPortal = isExitPortal;
        }

        public String getCharacterEncoding() {
            // NOTE: This is used only by ResultStoreWriter as of 8/12/03,
            // and used only to compute the size of the resulting byte
            // stream, size which is then set but not used when working with
            // portlets.
            return CachedSerializer.DEFAULT_ENCODING;
        }

        public boolean isCommitted() {
            return false;
        }

        public void reset() {
            // NOP
        }

        public boolean isRedirect() {
            return redirectPathInfo != null;
        }

        public boolean hasContent() {
            return byteStream != null || stringBuilderWriter != null;
        }

        public void setTitle(String title) {
            this.title = title;
        }

        public String getTitle() {
            return title;
        }
    }

    public Object getNativeRequest() {
        return portletRequest;
    }

    public Object getNativeResponse() {
        return mimeResponse;
    }

    public ExternalContext.Request getRequest() {
        if (request == null)
            request = new Request();
        return request;
    }

    public Response getResponse() {
        if (response == null)
            response = new BufferedResponse(pipelineContext, request);
        return response;
    }

    public ExternalContext.Session getSession(boolean create) {
        if (session == null) {
            PortletSession nativeSession = portletRequest.getPortletSession(create);
            if (nativeSession != null)
                session = new Session(nativeSession);
        }
        return session;
    }

    public ExternalContext.Application getApplication() {
        if (portletContext == null && portletRequest.getPortletSession() != null)
          portletContext = portletRequest.getPortletSession().getPortletContext();
        if (portletContext != null)
          application = new Application(portletContext);

        return application;
    }

    public String getStartLoggerString() {
        return getRequest().getRequestPath() + " - Received request";
    }

    public String getEndLoggerString() {
        return getRequest().getRequestPath();
    }

    public RequestDispatcher getRequestDispatcher(String path, boolean isContextRelative) {
        return new PortletToExternalContextRequestDispatcherWrapper(portletContext.getRequestDispatcher(path));
    }

    /**
     * Present a view of the HttpServletRequest properties as a Map.
     */
    public static class RequestMap extends AttributesToMap<Object> {
        public RequestMap(final PortletRequest portletRequest) {
            super(new Attributeable<Object>() {
                public Object getAttribute(String s) {
                    return portletRequest.getAttribute(s);
                }

                public Enumeration<String> getAttributeNames() {
                    return portletRequest.getAttributeNames();
                }

                public void removeAttribute(String s) {
                    portletRequest.removeAttribute(s);
                }

                public void setAttribute(String s, Object o) {
                    portletRequest.setAttribute(s, o);
                }
            });
        }
    }

    /**
     * Present a view of the HttpSession properties as a Map.
     */
    public static class SessionMap extends AttributesToMap<Object> {
        public SessionMap(final PortletSession portletSession, final int scope) {
            super(new Attributeable<Object>() {
                public Object getAttribute(String s) {
                    return portletSession.getAttribute(s, scope);
                }

                public Enumeration<String> getAttributeNames() {
                    return portletSession.getAttributeNames(scope);
                }

                public void removeAttribute(String s) {
                    portletSession.removeAttribute(s, scope);
                }

                public void setAttribute(String s, Object o) {
                    portletSession.setAttribute(s, o, scope);
                }
            });
        }
    }

    public String rewriteServiceURL(String urlString, int rewriteMode) {
        return URLRewriterUtils.rewriteServiceURL(getRequest(), urlString, rewriteMode);
    }
}

File: src/main/scala/org/orbeon/oxf/portlet/Portlet2ExternalContext.scala
/**
  * Copyright (C) 2010 Orbeon, Inc.
  *
  * This program is free software; you can redistribute it and/or modify it under the terms of the
  * GNU Lesser General Public License as published by the Free Software Foundation; either version
  * 2.1 of the License, or (at your option) any later version.
  *
  * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
  * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
  * See the GNU Lesser General Public License for more details.
  *
  * The full text of the license is available at http://www.gnu.org/copyleft/lesser.html
  */
package org.orbeon.oxf.portlet

import java.io._
import java.security.Principal
import java.util.Locale
import java.{util ⇒ ju}
import javax.portlet._

import org.orbeon.oxf.common.OXFException
import org.orbeon.oxf.externalcontext.{PortletToExternalContextRequestDispatcherWrapper, WSRPURLRewriter}
import org.orbeon.oxf.http.Headers
import org.orbeon.oxf.pipeline.api.{ExternalContext, PipelineContext}
import org.orbeon.oxf.util.CollectionUtils._
import org.orbeon.oxf.util.CoreUtils._
import org.orbeon.oxf.util.StringUtils._
import org.orbeon.oxf.util._
import org.orbeon.oxf.webapp.{HttpStatusCodeException, SessionListeners, WebAppContext, WebAppRequest}

import scala.collection.JavaConverters._

/*
 * Portlet-specific implementation of ExternalContext.
 */
object Portlet2ExternalContext {

  class BufferedResponseImpl(request: ExternalContext.Request)
    extends ExternalContext.Response {

    private val urlRewriter = new WSRPURLRewriter(
      getPathMatchers     = URLRewriterUtils.getPathMatchersCallable,
      request             = request,
      wsrpEncodeResources = URLRewriterUtils.isWSRPEncodeResources
    )

    // Response state
    private var contentType          : String  = null
    private var redirectLocation     : String  = null
    private var redirectIsExitPortal : Boolean = false
    private var title                : String  = null

    private lazy val streams = {
      val stringBuilderWriter = new StringBuilderWriter
      val printWriter         = new PrintWriter(stringBuilderWriter)
      val outputStream        = new ByteArrayOutputStream

      (stringBuilderWriter, printWriter, outputStream)
    }

    // Not handled right now or because it doesn't make sense
    def setContentLength(len: Int)                            = ()
    def checkIfModifiedSince(lastModified: Long)              = true
    def sendError(code: Int)                                  = throw new OXFException(s"Error while processing request: $code")
    def setPageCaching(lastModified: Long)                    = ()
    def setResourceCaching(lastModified: Long, expires: Long) = ()
    def setHeader(name: String, value: String)                = ()
    def addHeader(name: String, value: String)                = ()
    def reset()                                               = () // NOTE: We could implement this if needed.

    // We are always buffering
    def isCommitted                                           = false

    def setStatus(status: Int): Unit = {
      // Test error
      if (status == ExternalContext.SC_NOT_FOUND) {
        throw HttpStatusCodeException(ExternalContext.SC_NOT_FOUND)
      } else if (status >= 400) {
        // Ignore
      }
      // TODO: How to handle NOT_MODIFIED?
    }

    def rewriteActionURL(urlString: String): String =
      urlRewriter.rewriteActionURL(urlString)

    def rewriteRenderURL(urlString: String): String =
      urlRewriter.rewriteRenderURL(urlString)

    def rewriteActionURL(urlString: String, portletMode: String, windowState: String): String =
      urlRewriter.rewriteActionURL(urlString, portletMode, windowState)

    def rewriteRenderURL(urlString: String, portletMode: String, windowState: String): String =
      urlRewriter.rewriteRenderURL(urlString, portletMode, windowState)

    def rewriteResourceURL(urlString: String, rewriteMode: Int): String =
      urlRewriter.rewriteResourceURL(urlString, rewriteMode)

    def getNamespacePrefix: String = urlRewriter.getNamespacePrefix
    def getNativeResponse: AnyRef  = throw new NotImplementedError

    def setContentType(contentType: String) = this.contentType = NetUtils.getContentTypeMediaType(contentType)
    def setTitle(title: String)             = this.title = title

    def getWriter: PrintWriter         = streams._2
    def getOutputStream: OutputStream  = streams._3

    def sendRedirect(
      location     : String,
      isServerSide : Boolean, // ignored for portlets
      isExitPortal : Boolean  // if this is true, the redirect will exit the portal
    ): Unit = {
      this.redirectLocation     = location
      this.redirectIsExitPortal = isExitPortal
    }

    // NOTE: This is used only by ResultStoreWriter as of 8/12/03,
    // and used only to compute the size of the resulting byte
    // stream, size which is then set but not used when working with
    // portlets.
    def getCharacterEncoding: String = ExternalContext.DefaultHeaderEncoding

    def getContentType         : String  = contentType
    def getRedirectLocation    : String  = redirectLocation
    def isRedirectIsExitPortal : Boolean = redirectIsExitPortal
    def isRedirect             : Boolean = redirectLocation ne null
    def getTitle               : String  = title

    def getBytes: Array[Byte] =
      if (streams._1.getBuilder.length > 0)
        streams._1.toString.getBytes(getContentType) // TODO: what if content type not specified?
      else if (streams._3.size > 0)
        streams._3.toByteArray
      else
        Array.empty[Byte]
  }

  // Present a view of the HttpServletRequest properties as a Map.
  class RequestMap(val portletRequest: PortletRequest) extends AttributesToMap[AnyRef](
    new AttributesToMap.Attributeable[AnyRef] {
      def getAttribute(s: String): AnyRef           = portletRequest.getAttribute(s)
      def getAttributeNames: ju.Enumeration[String] = portletRequest.getAttributeNames
      def removeAttribute(s: String)                = portletRequest.removeAttribute(s)
      def setAttribute(s: String, o: AnyRef)        = portletRequest.setAttribute(s, o)
    }
  )

  // Present a view of the HttpSession properties as a Map.
  class SessionMap(val portletSession: PortletSession, val scope: Int) extends AttributesToMap[AnyRef](
    new AttributesToMap.Attributeable[AnyRef]() {
      def getAttribute(s: String): AnyRef           = portletSession.getAttribute(s, scope)
      def getAttributeNames: ju.Enumeration[String] = portletSession.getAttributeNames(scope)
      def removeAttribute(s: String)                = portletSession.removeAttribute(s, scope)
      def setAttribute(s: String, o: AnyRef)        = portletSession.setAttribute(s, o, scope)
    }
  )
}

class Portlet2ExternalContext(
  val pipelineContext : PipelineContext,
  val webAppContext   : WebAppContext,
  var portletRequest  : PortletRequest,
  val amendRequest    : Boolean
) extends ExternalContext {

  import Portlet2ExternalContext._

  private val clientDataRequestOpt = collectByErasedType[ClientDataRequest](portletRequest)

  def getWebAppContext: WebAppContext = webAppContext

  private class RequestImpl extends ExternalContext.Request with WebAppRequest {

    def getContainerType          : String            = "portlet"
    def getContainerNamespace     : String            = getResponse.getNamespacePrefix
    def getRequestPath            : String            = getPathInfo

    // Delegate to underlying request
    def getContextPath            : String            = portletRequest.getContextPath
    def getAuthType               : String            = portletRequest.getAuthType
    def isSecure                  : Boolean           = portletRequest.isSecure
    def getRequestedSessionId     : String            = portletRequest.getRequestedSessionId
    def getScheme                 : String            = portletRequest.getScheme
    def getServerName             : String            = portletRequest.getServerName
    def getServerPort             : Int               = portletRequest.getServerPort
    def getLocale                 : Locale            = portletRequest.getLocale
    def getLocales                : ju.Enumeration[_] = portletRequest.getLocales
    def isRequestedSessionIdValid : Boolean           = portletRequest.isRequestedSessionIdValid
    def getUserPrincipal          : Principal         = portletRequest.getUserPrincipal
    def getNativeRequest          : AnyRef            = portletRequest
    def getPortletMode            : String            = portletRequest.getPortletMode.toString
    def getWindowState            : String            = portletRequest.getWindowState.toString

    // Delegate to underlying request with data
    def getCharacterEncoding      : String            = clientDataRequestOpt map (_.getCharacterEncoding)  orNull
    def getContentLength          : Int               = clientDataRequestOpt map (_.getContentLength)      getOrElse -1
    def getContentType            : String            = clientDataRequestOpt map (_.getContentType)        orNull
    def getMethod                 : String            = clientDataRequestOpt map (_.getMethod)             getOrElse "GET"
    def getReader                 : Reader            = clientDataRequestOpt map (_.getReader)             orNull
    def getInputStream            : InputStream       = clientDataRequestOpt map (_.getPortletInputStream) orNull

    // Not available or not implemented
    def getRemoteAddr             : String            = null
    def getPathTranslated         : String            = null
    def getProtocol               : String            = null
    def getQueryString            : String            = null
    def getRequestURI             : String            = null
    def getRequestURL             : String            = null
    def getServletPath            : String            = null
    def getRemoteHost             : String            = null

    lazy val getPathInfo: String = {
      // Use the resource id if we are a ResourceRequest
      // In that case, remove the query string part of the resource id, that's handled by getParameterMap()
      val rawResult =
        portletRequest match {
          case rr: ResourceRequest ⇒ PathUtils.splitQuery(rr.getResourceID)._1
          case _                   ⇒ portletRequest.getParameter(WSRPURLRewriter.PathParameterName)
        }

      PathUtils.appendStartingSlash(rawResult.trimAllToEmpty)
    }

    protected lazy val headerValuesMap = {

      // NOTE: Not sure we should even pass these properties as "headers"
      // Example of property: javax.portlet.markup.head.element.support = true

      val propertiesIt =
        for (name ← portletRequest.getPropertyNames.asScala)
          yield name.toLowerCase → StringConversions.stringEnumerationToArray(portletRequest.getProperties(name))

      // PLT.11.1.5 Request Properties: "client request HTTP headers may not be always available. Portlets
      // should not rely on the presence of headers to function properly. The PortletRequest interface
      // provides specific methods to access information normally available as HTTP headers: content-length,
      // content-type, accept-language."
      // NOTE: It seems like while Liferay 5 was making headers available, Liferay 6 doesn't anymore.

      val headersIt =
        clientDataRequestOpt.iterator flatMap { cdr ⇒
          ((cdr.getContentType ne null) iterator (Headers.ContentTypeLower   → Array(cdr.getContentType))) ++
          ((cdr.getContentLength != -1) iterator (Headers.ContentLengthLower → Array(cdr.getContentLength.toString)))
        }

      propertiesIt ++ headersIt toMap
    }

    def getHeaderValuesMap: ju.Map[String, Array[String]] = headerValuesMap.asJava

    def getSession(create: Boolean): ExternalContext.Session =
      Portlet2ExternalContext.this.getSession(create)

    def sessionInvalidate(): Unit = {
      val session = portletRequest.getPortletSession(false)
      if (session ne null)
        session.invalidate()
    }

    lazy val getParameterMap: ju.Map[String, Array[AnyRef]] =
      if ((getContentType ne null) && getContentType.startsWith("multipart/form-data")) {
        Multipart.jGetParameterMapMultipart(pipelineContext, getRequest, ExternalContext.DefaultFormCharsetDefault)
      } else {
        portletRequest match {
          case rr: ResourceRequest ⇒
            // We encoded query parameters directly into the resource id in this case
            val (_, queryString) = PathUtils.splitQueryDecodeParams(rr.getResourceID)
            CollectionUtils.combineValues[String, AnyRef, Array](queryString).toMap.asJava
          case _ ⇒
            // Use native request parameters
            // Filter out `PathParameterName`, make values `Array[AnyRef]` (not great), and make immutable `Map`.
            val filteredParams =
              portletRequest.getParameterMap.asScala collect {
                case pair @ (k, v) if k != WSRPURLRewriter.PathParameterName ⇒ (k, v.toArray[AnyRef])
              }
            filteredParams.asJava
        }
      }

    lazy val getAttributesMap: ju.Map[String, AnyRef] =
      new RequestMap(portletRequest)
  }

  private class SessionImpl(portletSession: PortletSession)
    extends ExternalContext.Session {

    import PortletSession._

    import SessionListeners._

    // Delegate
    def getCreationTime                       = portletSession.getCreationTime
    def getId                                 = portletSession.getId
    def getLastAccessedTime                   = portletSession.getLastAccessedTime
    def getMaxInactiveInterval                = portletSession.getMaxInactiveInterval
    def invalidate()                          = portletSession.invalidate()
    def isNew                                 = portletSession.isNew
    def setMaxInactiveInterval(interval: Int) = portletSession.setMaxInactiveInterval(interval)

    lazy val applicationAttributesMap = new SessionMap(portletSession, APPLICATION_SCOPE)
    lazy val portletAttributesMap     = new SessionMap(portletSession, PORTLET_SCOPE)

    def getAttributesMap: ju.Map[String, AnyRef] = portletAttributesMap

    def getAttributesMap(scope: Int): ju.Map[String, AnyRef] = scope match {
      case APPLICATION_SCOPE ⇒ applicationAttributesMap
      case PORTLET_SCOPE     ⇒ portletAttributesMap
      case _                 ⇒ throw new IllegalArgumentException
    }

    def addListener(sessionListener: ExternalContext.Session.SessionListener): Unit =
      portletSession.getAttribute(SessionListenersKey, APPLICATION_SCOPE) match {
        case listeners: SessionListeners ⇒ listeners.addListener(sessionListener)
        case _ ⇒
          throw new IllegalStateException(
            "`SessionListeners` object not found in session. `OrbeonSessionListener` might be missing from web.xml."
          )
      }

    def removeListener(sessionListener: ExternalContext.Session.SessionListener): Unit =
      portletSession.getAttribute(SessionListenersKey, APPLICATION_SCOPE) match {
        case listeners: SessionListeners ⇒ listeners.removeListener(sessionListener)
        case _ ⇒
      }
  }

  lazy val getRequest  : ExternalContext.Request  = new RequestImpl
  lazy val getResponse : ExternalContext.Response = new BufferedResponseImpl(getRequest)

  private var sessionImplOpt: Option[SessionImpl] = None

  def getSession(create: Boolean): ExternalContext.Session =
    sessionImplOpt getOrElse {
      val nativeSession = portletRequest.getPortletSession(create)

      if (nativeSession ne null) {
        val newSessionImpl = new SessionImpl(nativeSession)
        sessionImplOpt = Some(newSessionImpl)
        newSessionImpl
      } else
        null
    }

  def getStartLoggerString: String = getRequest.getRequestPath + " - Received request"
  def getEndLoggerString  : String = getRequest.getRequestPath

  def getRequestDispatcher(path: String, isContextRelative: Boolean): ExternalContext.RequestDispatcher =
    new PortletToExternalContextRequestDispatcherWrapper(
      webAppContext.getNativeContext.asInstanceOf[PortletContext].getRequestDispatcher(path)
    )
}